[rfc][be] static assert that nccl version is >= 2.7

c-p-i-o opened the pull request:
Summary:
Static assert that NCCL VERSION is greater than 2.4.
This is in preparation of enabling error checking by default in PyTorch library and removal of some macros.
This is in PR #141914. 
The rationale behind this version is:
1. 2.4 released ~2 years ago so it's unlikely that someone is still using the old library.
2. Enabling error checking is benefitial to the community as it helps debug subtle bugs in production environments.

Test Plan: unit tests

Differential Revision: D66737055




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -39,6 +39,11 @@ constexpr int64_t kCommInitBusyWaitMillis = 2;
 #define NCCL_REMOTE_ERROR
 #endif
 
+#define HAS_NCCL_MAJOR_MINOR (defined(NCCL_MAJOR) && defined(NCCL_MINOR))
+#if HAS_NCCL_MAJOR_MINOR
+static_assert((NCCL_MAJOR == 2 && NCCL_MINOR >= 4) ||   (NCCL_MAJOR > 2), "NCCL version must be 2.4 or later");                                                            
+#endif  // HAS_NCCL_MAJOR_MINOR
+
```
To work around the CI's concern, we can directly static_assert. I should be fair to assume both `NCCL_MAJOR` and `NCCL_MINOR` are defined when nccl.h is included.

c-p-i-o replied:
I was concerned about the case where  someone builds PyTorch but with no NCCL library. 
So this file would unnecessarily assert for this case.
If you are fairly sure that NCCL is always required to build PyTorch, I can add the static_assert unconditionally?

kwen2501 reviewed (approved):
LGTM.

c-p-i-o committed:
[rfc][be] static assert that nccl version is >= 2.7 (#142023)

Summary:

Static assert that NCCL VERSION is greater than 2.7.
This is in prepration of enabling error checking by default in PyTorch library.
The rationale behind this version is:
1. 2.7 released ~6 years ago so it's unlikely that someone is still using the old library.
2. Enabling error checking is benefitial to the community as it helps debug subtle bugs in production environments. So this change is just in preparation for some code cleanup.

Test Plan: unit tests

Reviewed By: kwen2501

Differential Revision: D66737055
